File: ios-lister/CloudStoreLister/Logic/ListExportLogic.swift
import Foundation

enum ListExportLogic {
    struct Summary: Equatable {
        let itemCount: Int
        let totalPullCount: Int
    }

    static func summary(for items: [InventoryListItem]) -> Summary {
        Summary(
            itemCount: items.count,
            totalPullCount: items.reduce(0) { $0 + $1.pullCount }
        )
    }

    static func buildCSVString(for items: [InventoryListItem]) -> String {
        var csv = "Name,Type,Product ID,Barcode,Price,Stock,Pull Count\n"
        for item in items {
            csv += [
                csvQuoted(item.name),
                csvQuoted(item.productType ?? ""),
                csvQuoted("\(item.productId)"),
                csvQuoted(item.barcode ?? ""),
                csvQuoted(item.priceLabel),
                csvQuoted(item.stockLabel),
                "\(item.pullCount)",
            ].joined(separator: ",")
            csv += "\n"
        }
        return csv
    }

    static func writeCSVFile(items: [InventoryListItem], listName: String) -> URL? {
        let safeName = sanitizedFileName(listName)
        let timestamp = Int(Date().timeIntervalSince1970)
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(safeName)_\(timestamp).csv")
        do {
            try buildCSVString(for: items).write(to: url, atomically: true, encoding: .utf8)
            return url
        } catch {
            return nil
        }
    }

    private static func csvQuoted(_ value: String) -> String {
        "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
    }

    private static func sanitizedFileName(_ name: String) -> String {
        let noSpaces = name.replacingOccurrences(of: " ", with: "_")
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-_"))
        let cleaned = noSpaces
            .unicodeScalars
            .filter { allowed.contains($0) }
            .map { String($0) }
            .joined()
        return cleaned.isEmpty ? "list" : cleaned
    }
}

File: android-lister/app/src/main/java/com/cloudstore/lister/domain/ListExportLogic.kt
package com.cloudstore.lister.domain

import com.cloudstore.lister.data.InventoryListItem
import java.io.File

object ListExportLogic {
    data class Summary(val itemCount: Int, val totalPullCount: Int)

    fun summary(items: List<InventoryListItem>): Summary {
        return Summary(
            itemCount = items.size,
            totalPullCount = items.sumOf { it.pullCount },
        )
    }

    fun buildCsv(items: List<InventoryListItem>): String {
        val header = "Name,Type,Product ID,Barcode,Price,Stock,Pull Count\n"
        val rows = items.joinToString("\n") { item ->
            listOf(
                csvQuoted(item.name),
                csvQuoted(item.productType.orEmpty()),
                csvQuoted(item.productId.toString()),
                csvQuoted(item.barcode.orEmpty()),
                csvQuoted(item.priceLabel),
                csvQuoted(item.stockLabel),
                item.pullCount.toString(),
            ).joinToString(",")
        }
        return header + rows + if (rows.isEmpty()) "" else "\n"
    }

    fun writeCsvFile(cacheDir: File, listName: String, items: List<InventoryListItem>): File? {
        return runCatching {
            val safeName = sanitizedFileName(listName)
            val file = File(cacheDir, "${safeName}_${System.currentTimeMillis()}.csv")
            file.writeText(buildCsv(items))
            file
        }.getOrNull()
    }

    fun sanitizedFileName(name: String): String {
        val noSpaces = name.replace(' ', '_')
        val cleaned = noSpaces.filter { it.isLetterOrDigit() || it == '-' || it == '_' }
        return cleaned.ifEmpty { "list" }
    }

    private fun csvQuoted(value: String): String {
        return "\"${value.replace("\"", "\"\"")}\""
    }
}

object ListQueryLogic {
    fun resultListName(sourceName: String, customName: String): String {
        val trimmed = customName.trim()
        return trimmed.ifEmpty { "$sourceName-BQ" }
    }

    fun setItems(items: List<InventoryListItem>, forListId: String, lists: List<com.cloudstore.lister.data.InventoryNamedList>): List<com.cloudstore.lister.data.InventoryNamedList> {
        val index = lists.indexOfFirst { it.id == forListId }
        if (index < 0) return lists
        return lists.toMutableList().also { it[index] = it[index].copy(items = items) }
    }
}
